src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -2999,23 +2999,53 @@ private void VisitStatementsWithLocalFunctions(BoundBlock block)
             if (!TrackingRegions && !block.LocalFunctions.IsDefaultOrEmpty)
             {
                 // First visit everything else
+                var queue1 = ArrayBuilder<BoundLocalFunctionStatement>.GetInstance();
                 foreach (var stmt in block.Statements)
                 {
-                    if (stmt.Kind != BoundKind.LocalFunctionStatement)
+                    if (stmt is BoundLocalFunctionStatement localFunc)
+                    {
+                        queue1.Add(localFunc);
+                    }
+                    else
                     {
                         VisitStatement(stmt);
                     }
                 }
 
-                // Now visit the local function bodies
-                foreach (var stmt in block.Statements)
+                // Now visit the local function bodies.
+                // We first visit those we have seen called (hence we know their starting state),
+                // repeating this while seeing more call sites inside other local functions,
+                // and only then we visit the completely unreachable bodies.
+                // This avoids unnecessary passes and incorrect starting states.
+                var queue2 = ArrayBuilder<BoundLocalFunctionStatement>.GetInstance();
+                for (int c = 0; c != LocalFuncUsageCount;)
```

Comment:
>LocalFuncUsageCount [](http://example.com/codeflow?start=37&length=19)

I am not sure why value of ```LocalFuncUsageCount``` is significant. It looks like it could be incremented by analysis for local functions declared in nested blocks, or in blocks preceding the current block. That would mean that its value might say nothing about usages of local functions in the current block.

Reply:
Btw, I think this might still be useful heuristic to avoid unnecessary iterations of the algorithm. I.e., if we kept this global `LocalFuncUsageCount` and it didn't increment, we would know that we don't have to iterate the first inner `for` loop and can go to visiting the first unused function below.